File: client/services/learn.ts
/**
 * Learning & Gamification API service
 */
import api from './api';

// Types

export interface SubjectStudent {
  id: string;
  name: string;
  code: string;
  description?: string;
  teacher_name?: string;
  total_topics: number;
  total_questions: number;
  is_enrolled: boolean;
  enrollment_status?: string; // pending, approved, rejected, or null
  mastery: number;
  xp_earned: number;
}

export interface Enrollment {
  id: string;
  student_id: string;
  subject_id: string;
  enrolled_at: string;
  is_active: boolean;
  status: string;
  subject_name?: string;
  subject_code?: string;
}

export interface LessonQuestion {
  id: string;
  question_text: string;
  question_type?: string;
  options?: string[];
  difficulty_level?: string;
  bloom_taxonomy_level?: string;
  marks?: number;
  topic_id?: string;
}

export interface LessonData {
  subject_id: string;
  topic_id?: string;
  topic_name?: string;
  difficulty: string;
  questions: LessonQuestion[];
  total_questions: number;
  hearts_remaining: number;
}

export interface AnswerSubmission {
  question_id: string;
  selected_answer: string;
  time_taken_seconds?: number;
}

export interface AnswerResult {
  question_id: string;
  is_correct: boolean;
  correct_answer: string;
  xp_earned: number;
  explanation?: string;
}

export interface LessonResult {
  score: number;
  total_marks: number;
  total_questions: number;
  correct_answers: number;
  xp_earned: number;
  streak_maintained: boolean;
  new_streak_count: number;
  hearts_remaining: number;
  mastery_change: number;
  new_mastery: number;
  level_up: boolean;
  new_level: number;
  results: AnswerResult[];
  accuracy: number;
  tutor_feedback?: string;
}

export interface StudentProgress {
  id: string;
  student_id: string;
  subject_id: string;
  topic_id?: string;
  topic_name?: string;
  topic_mastery: number;
  xp_earned: number;
  current_level: number;
  accuracy_percentage: number;
  questions_attempted: number;
  questions_correct: number;
  current_difficulty: string;
}

export interface GamificationProfile {
  user_id: string;
  username: string;
  full_name?: string;
  avatar_url?: string;
  role: string;
  xp_total: number;
  streak_count: number;
  hearts: number;
  current_level: number;
  badges: string[];
  subjects_enrolled: number;
  total_lessons_completed: number;
  total_questions_answered: number;
  overall_accuracy: number;
}

export interface LeaderboardEntry {
  rank: number;
  user_id: string;
  username: string;
  full_name?: string;
  avatar_url?: string;
  xp_total: number;
  streak_count: number;
  level: number;
}

export interface LeaderboardData {
  entries: LeaderboardEntry[];
  total_students: number;
  current_user_rank?: number;
}

export interface TestHistoryEntry {
  id: string;
  student_id: string;
  subject_id: string;
  topic_id?: string;
  score: number;
  total_marks: number;
  total_questions: number;
  correct_answers: number;
  xp_earned: number;
  time_taken_seconds?: number;
  difficulty: string;
  created_at: string;
}

export interface DailyActivityEntry {
  id: string;
  student_id: string;
  activity_date: string;
  xp_earned: number;
  questions_answered: number;
  correct_answers: number;
  time_spent_seconds: number;
  lessons_completed: number;
}

export interface HeartsData {
  hearts: number;
  max_hearts: number;
}

// API Methods

const learnService = {
  // Subjects
  async getAvailableSubjects(): Promise<SubjectStudent[]> {
    const { data } = await api.get('/learn/subjects');
    return data;
  },

  // Topics for enrolled subjects (student-accessible)
  async getTopics(subjectId: string): Promise<{ topics: { id: string; name: string; description?: string; order_index: number; has_syllabus: boolean; syllabus_content?: string; total_questions: number }[] }> {
    const { data } = await api.get(`/learn/topics/${subjectId}`);
    return data;
  },

  // Student-accessible reference materials (only is_public=true docs)
  async getStudentReferences(subjectId: string): Promise<{ reference_books: any[]; template_papers: any[]; reference_questions: any[] }> {
    const { data } = await api.get(`/learn/references/${subjectId}`);
    return data;
  },

  // Enrollment
  async enrollInSubject(subjectId: string): Promise<Enrollment> {
    const { data } = await api.post('/learn/enroll', { subject_id: subjectId });
    return data;
  },

  async getEnrollments(): Promise<Enrollment[]> {
    const { data } = await api.get('/learn/enrollments');
    return data;
  },

  async getAllEnrollments(): Promise<Enrollment[]> {
    const { data } = await api.get('/learn/enrollments/all');
    return data;
  },

  // Lessons
  async getLesson(subjectId: string, topicId?: string, count: number = 10): Promise<LessonData> {
    const params: Record<string, string | number> = { count };
    if (topicId) params.topic_id = topicId;
    const { data } = await api.get(`/learn/lesson/${subjectId}`, { params });
    return data;
  },

  async submitLesson(submission: {
    subject_id: string;
    topic_id?: string;
    answers: AnswerSubmission[];
    total_time_seconds?: number;
  }): Promise<LessonResult> {
    const { data } = await api.post('/learn/lesson/submit', submission);
    return data;
  },

  // Progress
  async getSubjectProgress(subjectId: string): Promise<StudentProgress[]> {
    const { data } = await api.get(`/learn/progress/${subjectId}`);
    return data;
  },

  // Profile
  async getProfile(): Promise<GamificationProfile> {
    const { data } = await api.get('/learn/profile');
    return data;
  },

  // Leaderboard
  async getLeaderboard(limit: number = 20, subjectId?: string): Promise<LeaderboardData> {
    const params: Record<string, string | number> = { limit };
    if (subjectId) params.subject_id = subjectId;
    const { data } = await api.get('/learn/leaderboard', { params });
    return data;
  },

  // History
  async getTestHistory(subjectId?: string, limit: number = 20): Promise<TestHistoryEntry[]> {
    const params: Record<string, string | number> = { limit };
    if (subjectId) params.subject_id = subjectId;
    const { data } = await api.get('/learn/history', { params });
    return data;
  },

  // Activity
  async getDailyActivity(days: number = 30): Promise<DailyActivityEntry[]> {
    const { data } = await api.get('/learn/activity', { params: { days } });
    return data;
  },

  // Hearts
  async getHearts(): Promise<HeartsData> {
    const { data } = await api.get('/learn/hearts');
    return data;
  },
};

export default learnService;

File: backend/app/api/v1/endpoints/learn.py
"""
Learning & Gamification API endpoints for students.
"""

from typing import Optional, List
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.api.v1.deps import get_current_active_user
from app.models.user import User
from app.models.subject import Topic
from app.models.gamification import Enrollment
from app.models.document import Document
from app.services.gamification_service import GamificationService
from app.schemas.gamification import (
    EnrollmentCreate,
    EnrollmentResponse,
    StudentProgressResponse,
    LessonResponse,
    LessonSubmission,
    LessonResult,
    TestHistoryResponse,
    DailyActivityResponse,
    GamificationProfile,
    LeaderboardResponse,
    SubjectListStudent,
)

router = APIRouter()


# --- Subjects for students ---

@router.get("/subjects", response_model=list[SubjectListStudent])
async def list_available_subjects(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """List all published subjects with enrollment status."""
    service = GamificationService(db)
    return await service.get_available_subjects(current_user.id)


@router.get("/topics/{subject_id}")
async def list_student_topics(
    subject_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """List topics for a subject that the student is enrolled in.
    Returns topic names, IDs, syllabus content, and question counts.
    """
    # Verify enrollment
    enrollment = await db.execute(
        select(Enrollment).where(
            Enrollment.student_id == current_user.id,
            Enrollment.subject_id == subject_id,
            Enrollment.status == "approved",
        )
    )
    if not enrollment.scalar_one_or_none():
        raise HTTPException(status_code=403, detail="Not enrolled in this subject")

    # Get topics
    result = await db.execute(
        select(Topic)
        .where(Topic.subject_id == subject_id)
        .order_by(Topic.order_index)
    )
    topics = result.scalars().all()

    return {
        "topics": [
            {
                "id": str(t.id),
                "name": t.name,
                "description": t.description,
                "order_index": t.order_index,
                "has_syllabus": t.has_syllabus,
                "syllabus_content": t.syllabus_content if t.has_syllabus else None,
                "total_questions": t.total_questions,
            }
            for t in topics
        ]
    }


@router.get("/references/{subject_id}")
async def list_student_references(
    subject_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """List reference documents shared by the teacher for an enrolled subject.
    Only returns documents where is_public=True.
    """
    # Verify enrollment
    enrollment = await db.execute(
        select(Enrollment).where(
            Enrollment.student_id == current_user.id,
            Enrollment.subject_id == subject_id,
            Enrollment.status == "approved",
        )
    )
    if not enrollment.scalar_one_or_none():
        raise HTTPException(status_code=403, detail="Not enrolled in this subject")

    # Get public reference documents for this subject
    result = await db.execute(
        select(Document).where(
            Document.subject_id == subject_id,
            Document.is_public == True,
            Document.index_type.in_(["reference_book", "template_paper"]),
            Document.processing_status == "completed",
        ).order_by(Document.upload_timestamp.desc())
    )
    docs = result.scalars().all()

    doc_responses = [
        {
            "id": str(doc.id),
            "filename": doc.filename,
            "file_size_bytes": doc.file_size_bytes,
            "mime_type": doc.mime_type,
            "index_type": doc.index_type,
            "subject_id": str(doc.subject_id) if doc.subject_id else None,
            "processing_status": doc.processing_status,
            "upload_timestamp": doc.upload_timestamp.isoformat() if doc.upload_timestamp else None,
        }
        for doc in docs
    ]

    return {
        "reference_books": [d for d in doc_responses if d["index_type"] == "reference_book"],
        "template_papers": [d for d in doc_responses if d["index_type"] == "template_paper"],
        "reference_questions": [],
    }



@router.post("/enroll", response_model=EnrollmentResponse)
async def enroll_in_subject(
    data: EnrollmentCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Request enrollment in a subject (creates pending enrollment)."""
    service = GamificationService(db)
    try:
        enrollment = await service.enroll_student(current_user.id, data.subject_id)
        # Get subject info for response
        all_enrollments = await service.get_all_enrollments(current_user.id)
        for e in all_enrollments:
            if e["id"] == enrollment.id:
                return e
        return enrollment
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/enrollments", response_model=list[EnrollmentResponse])
async def list_enrollments(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """List current student's approved enrollments."""
    service = GamificationService(db)
    return await service.get_enrollments(current_user.id)


@router.get("/enrollments/all", response_model=list[EnrollmentResponse])
async def list_all_enrollments(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """List all enrollments (including pending/rejected)."""
    service = GamificationService(db)
    return await service.get_all_enrollments(current_user.id)


# --- Lessons ---

@router.get("/lesson/{subject_id}", response_model=LessonResponse)
async def get_lesson(
    subject_id: UUID,
    topic_id: Optional[UUID] = Query(None),
    count: int = Query(10, ge=1, le=50),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get a lesson (set of questions) for a subject/topic. Requires enrollment."""
    service = GamificationService(db)
    # Verify enrollment
    enrollments = await service.get_enrollments(current_user.id)
    if not any(e["subject_id"] == subject_id for e in enrollments):
        raise HTTPException(status_code=403, detail="Not enrolled in this subject")
    return await service.get_lesson_questions(current_user.id, subject_id, topic_id, count)


@router.post("/lesson/submit", response_model=LessonResult)
async def submit_lesson(
    submission: LessonSubmission,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Submit lesson answers and get results with XP/streak updates."""
    service = GamificationService(db)
    # Verify enrollment
    enrollments = await service.get_enrollments(current_user.id)
    if not any(e["subject_id"] == submission.subject_id for e in enrollments):
        raise HTTPException(status_code=403, detail="Not enrolled in this subject")
    try:
        return await service.submit_lesson(current_user.id, submission.model_dump())
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


# --- Progress ---

@router.get("/progress/{subject_id}", response_model=list[StudentProgressResponse])
async def get_subject_progress(
    subject_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get student progress for all topics in a subject."""
    service = GamificationService(db)
    return await service.get_subject_progress(current_user.id, subject_id)


# --- Profile & Gamification ---

@router.get("/profile", response_model=GamificationProfile)
async def get_gamification_profile(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get gamification profile (XP, streaks, badges, stats)."""
    service = GamificationService(db)
    try:
        return await service.get_profile(current_user.id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/leaderboard", response_model=LeaderboardResponse)
async def get_leaderboard(
    limit: int = Query(20, ge=1, le=100),
    subject_id: Optional[UUID] = Query(None, description="Filter leaderboard to students enrolled in this subject"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get XP leaderboard. Optionally filter by subject for class-wise ranking."""
    service = GamificationService(db)
    return await service.get_leaderboard(current_user.id, limit, subject_id)


# --- Test History ---

@router.get("/history", response_model=list[TestHistoryResponse])
async def get_test_history(
    subject_id: Optional[UUID] = Query(None),
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get test/lesson history."""
    service = GamificationService(db)
    return await service.get_test_history(current_user.id, subject_id, limit)


# --- Daily Activity ---

@router.get("/activity", response_model=list[DailyActivityResponse])
async def get_daily_activity(
    days: int = Query(30, ge=1, le=365),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get daily activity for streak tracking."""
    service = GamificationService(db)
    return await service.get_daily_activity(current_user.id, days)


# --- Hearts ---

@router.get("/hearts")
async def get_hearts(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get current heart count."""
    service = GamificationService(db)
    hearts = await service.get_hearts(current_user.id)
    return {"hearts": hearts, "max_hearts": 5}
